src/main/java/net/glowstone/util/compiler/EvalTask.java

```diff
@@ -0,0 +1,185 @@
+import javax.tools.JavaFileObject;
+import javax.tools.SimpleJavaFileObject;
+import javax.tools.StandardJavaFileManager;
+import javax.tools.ToolProvider;
+import net.glowstone.GlowServer;
+import net.glowstone.ServerProvider;
+import net.glowstone.util.ReflectionProcessor;
+import org.bukkit.ChatColor;
+
+/**
+ * A task for which a source string must be compiled and ran to complete.
+ */
+public class EvalTask implements Runnable {
+
+    private final boolean output;
+    private final String source;
+    private static Object lastOutput = null;
+    final Map<String, byte[]> classData = new HashMap<>();
+    private MapClassLoader classLoader;
+    JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
+
+    /**
+     * Create a new EvalTask.
+     *
+     * @param command The statement to evaluate.
+     * @param output Whether the statement returns an object.
+     */
+    public EvalTask(String command, boolean output) {
+        this.output = output;
+        String evaluatedCommand;
+        if (output) {
+            evaluatedCommand = command.substring(1);
+        } else {
+            evaluatedCommand = command;
+        }
+        if (compiler == null) {
+            GlowServer.logger.info("Glowstone is not running with a JDK and REPL is falling"
+                + "back to basic parsing using reflection.");
+            source = evaluatedCommand;
+        } else {
+            source = "import org.bukkit.*;\n"
+                + "import net.glowstone.*;\n"
+                + "public class REPLShell {\n"
+                + (output ? "public static Object run(Object last) {\n"
+                    : "public static void run(Object last) {\n") + evaluatedCommand + "\n}\n}\n";
+        }
+    }
+
+    /**
+     * Compiles and runs the source associated with this evaluation task.
+     */
+    @Override
+    public void run() {
+        if (compiler == null) {
+            ReflectionProcessor processor = new ReflectionProcessor(source,
+                ServerProvider.getServer());
+            Object result = processor.process();
+            GlowServer.logger.info(
+                ChatColor.GOLD + "Eval returned: " + (result == null ? ChatColor.RED
+                    + "<no value>"
+                    : ChatColor.AQUA + result.toString()));
+            return;
+        }
+        classLoader = new MapClassLoader();
+        ClassDataFileManager classDataFileManager = new ClassDataFileManager(compiler
+            .getStandardFileManager(null, null, null));
+
+        List<JavaFileObject> compilationUnit = new ArrayList<>();
+        compilationUnit.add(new JavaSource("REPLShell", source));
+
+        JavaCompiler.CompilationTask task = compiler.getTask(null, classDataFileManager,
+            null, null, null, compilationUnit);
+        if (task.call()) {
+            try {
+                Object returned = MethodInvocationUtils
+                    .invokeStaticMethod(getCompiledClass("REPLShell"), "run",
```

Comment:
Move the methods of MethodInvocationUtils to this class, unless and until another class uses them.